File: app/src/main/java/net/ddns/raspi_server/rezeptbuch/util/Favorite.java
package net.ddns.raspi_server.rezeptbuch.util;

import android.content.Context;
import android.content.SharedPreferences;

import net.ddns.raspi_server.rezeptbuch.Rezeptbuch;
import net.ddns.raspi_server.rezeptbuch.util.DataStructures.Recipe;
import net.ddns.raspi_server.rezeptbuch.util.db.RecipeDatabase;

import java.util.ArrayList;
import java.util.List;

public class Favorite {

  private static final String PREFERENCES = "FavoritePrefs";
  private static final String KEY_FAVORITE = "FavoriteKey";

  private final SharedPreferences mPreferences;
  private final RecipeDatabase mDB;

  private static class Holder {
    private static final Favorite INSTANCE = new Favorite();
  }

  public static Favorite getInstance() {
    return Holder.INSTANCE;
  }

  private Favorite() {
    Context context = Rezeptbuch.getContext();
    mPreferences = context.getSharedPreferences(PREFERENCES, Context
        .MODE_PRIVATE);
    mDB = new RecipeDatabase(context);
  }

  public List<Recipe> getRecipes() {
    List<Recipe> list = new ArrayList<>();
    String favoriteString = mPreferences.getString(KEY_FAVORITE, "");
    String[] strIDList = favoriteString.split(";");
    for (String str : strIDList)
      try {
        int id = Integer.parseInt(str);
        Recipe recipe = mDB.getRecipeById(id);
        if (recipe != null)
          list.add(recipe);
      } catch (NumberFormatException e) {
        e.printStackTrace();
      }
    return list;
  }

  public boolean contains(Recipe recipe) {
    return getRecipes().contains(recipe);
  }

  /**
   * @param recipe the recipe to toggle
   * @return true if the recipe was added, false if it was removed.
   */
  public boolean toggleRecipe(Recipe recipe) {
    boolean result = true;
    String favoriteString = mPreferences.getString(KEY_FAVORITE, "");

    int newItem = recipe._ID;
    String[] strIDArray = favoriteString.split(";");
    List<Integer> idList = new ArrayList<>();
    for (String str : strIDArray)
      try {
        idList.add(Integer.parseInt(str));
      } catch (NumberFormatException e) {
        e.printStackTrace();
      }

    for (int i = 0; i < idList.size(); i++) {
      if (idList.get(i) == newItem) {
        idList.remove(i);
        result = false;
        break;
      }
    }
    if (result)
      idList.add(0, newItem);
    favoriteString = "";
    for (Integer i : idList)
      favoriteString += i + ";";
    if (favoriteString.length() > 0)
      favoriteString = favoriteString.substring(0, favoriteString.length() - 1);
    mPreferences.edit().putString(KEY_FAVORITE, favoriteString).apply();
    return result;
  }
}

File: app/src/main/java/net/ddns/raspi_server/rezeptbuch/util/Favorite.kt
package net.ddns.raspi_server.rezeptbuch.util

import android.content.Context
import android.content.SharedPreferences

import net.ddns.raspi_server.rezeptbuch.Rezeptbuch
import net.ddns.raspi_server.rezeptbuch.util.DataStructures.Recipe
import net.ddns.raspi_server.rezeptbuch.util.db.RecipeDatabase

import java.util.ArrayList

object Favorite {

  private const val PREFERENCES = "FavoritePrefs"
  private const val KEY_FAVORITE = "FavoriteKey"

  private val mPreferences: SharedPreferences
  private val mDB: RecipeDatabase

  init {
    val context = Rezeptbuch.context
    mPreferences = context!!.getSharedPreferences(PREFERENCES, Context
            .MODE_PRIVATE)
    mDB = RecipeDatabase(context)
  }

  fun getRecipes(): List<Recipe> {
    val list = ArrayList<Recipe>()
    val favoriteString = mPreferences.getString(KEY_FAVORITE, "")
    val strIDList = favoriteString!!.split(";".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
    for (str in strIDList)
      try {
        val id = Integer.parseInt(str)
        val recipe = mDB.getRecipeById(id)
        if (recipe != null)
          list.add(recipe)
      } catch (e: NumberFormatException) {
        e.printStackTrace()
      }

    return list
  }

  operator fun contains(recipe: Recipe): Boolean {
    return getRecipes().contains(recipe)
  }

  /**
   * @param recipe the recipe to toggle
   * @return true if the recipe was added, false if it was removed.
   */
  fun toggleRecipe(recipe: Recipe): Boolean {
    var result = true
    var favoriteString = mPreferences.getString(KEY_FAVORITE, "")

    val newItem = recipe._ID
    val strIDArray = favoriteString!!.split(";".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
    val idList = ArrayList<Int>()
    for (str in strIDArray)
      try {
        idList.add(Integer.parseInt(str))
      } catch (e: NumberFormatException) {
        e.printStackTrace()
      }

    for (i in idList.indices) {
      if (idList[i] == newItem) {
        idList.removeAt(i)
        result = false
        break
      }
    }
    if (result)
      idList.add(0, newItem)
    favoriteString = ""
    for (i in idList)
      favoriteString += i.toString() + ";"
    if (favoriteString.length > 0)
      favoriteString = favoriteString.substring(0, favoriteString.length - 1)
    mPreferences.edit().putString(KEY_FAVORITE, favoriteString).apply()
    return result
  }
}
